source/extensions/filters/network/common/redis/redis_command_stats.cc

```diff
@@ -0,0 +1,107 @@
+#include "extensions/filters/network/common/redis/redis_command_stats.h"
+
+#include "extensions/filters/network/common/redis/supported_commands.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Common {
+namespace Redis {
+
+RedisCommandStats::RedisCommandStats(Stats::Scope& scope, const std::string& prefix, bool enabled)
+    : scope_(scope), stat_name_pool_(scope.symbolTable()), prefix_(stat_name_pool_.add(prefix)),
+      enabled_(enabled), upstream_rq_time_(stat_name_pool_.add("upstream_rq_time")),
+      latency_(stat_name_pool_.add("latency")), total_(stat_name_pool_.add("total")),
+      success_(stat_name_pool_.add("success")), error_(stat_name_pool_.add("error")),
+      unused_metric_(stat_name_pool_.add("unused")), null_metric_(stat_name_pool_.add("null")),
+      unknown_metric_(stat_name_pool_.add("unknown")) {
+  // Note: Even if this is disabled, we track the upstream_rq_time.
+  if (enabled_) {
+    // Create StatName for each Redis command. Note that we don't include Auth or Ping.
+    for (const std::string& command :
+         Extensions::NetworkFilters::Common::Redis::SupportedCommands::simpleCommands()) {
+      addCommandToPool(command);
+    }
+    for (const std::string& command :
+         Extensions::NetworkFilters::Common::Redis::SupportedCommands::evalCommands()) {
+      addCommandToPool(command);
+    }
+    for (const std::string& command : Extensions::NetworkFilters::Common::Redis::SupportedCommands::
+             hashMultipleSumResultCommands()) {
+      addCommandToPool(command);
+    }
+    addCommandToPool(Extensions::NetworkFilters::Common::Redis::SupportedCommands::mget());
+    addCommandToPool(Extensions::NetworkFilters::Common::Redis::SupportedCommands::mset());
+  }
+}
+
+void RedisCommandStats::addCommandToPool(const std::string& command_string) {
+  Stats::StatName command = stat_name_pool_.add(command_string);
+  stat_name_map_[command_string] = command;
+}
+
+Stats::Counter& RedisCommandStats::counter(const Stats::StatNameVec& stat_names) {
+  const Stats::SymbolTable::StoragePtr storage_ptr = scope_.symbolTable().join(stat_names);
```

Comment:
With my suggestion to have a single RedisCommandStats shared across all upstreams, I think we should remove scope_ from RedisCommandStats, and just have a symbol table. You can then pass in the scope to this function and to histogram() as that will vary depending on the upstream.

Reply:
Ok, interesting, I'll see if I can put that together quick and test it in my VM. As I understand it, we'll:
1) Get the scope from the client/connection pool at counter/histogram time
2) Create a symbol table once in config, off the scope we use to create our per-upstream scope, ie `Stats::SymbolTable& symbol_table = context.scope().symbolTable();`

Leaving our `counter()` functions looking like so

```
Stats::Counter& RedisCommandStats::counter(const Stats::StatNameVec& stat_names) {
  const Stats::SymbolTable::StoragePtr storage_ptr = scope_.symbolTable().join(stat_names);
  Stats::StatName full_stat_name = Stats::StatName(storage_ptr.get());
  return scope_.counterFromStatName(full_stat_name);
}
```

```
Stats::Counter& RedisCommandStats::counter(Stats::Scope& scope, const Stats::StatNameVec& stat_names) {
  const Stats::SymbolTable::StoragePtr storage_ptr = symbol_table_.join(stat_names);
  Stats::StatName full_stat_name = Stats::StatName(storage_ptr.get());
  return scope.counterFromStatName(full_stat_name);
}
```